src/riak_kv_qry_compiler.erl

```diff
@@ -1097,9 +1110,378 @@ modify_where_key(TupleList, Field, NewVal) ->
     lists:keyreplace(Field, 1, TupleList, {Field, FieldType, NewVal}).
 
+-record(filtercheck, {name,'=','>','>=','<','<='}).
+
+check_where_clause_is_possible(DDL, WhereProps) ->
+    Filter1 = proplists:get_value(filter, WhereProps),
+    {Filter2, FilterChecks} = riak_ql_ddl:mapfold_where_tree(
+        fun (_, and_, Acc) ->
+                {ok, Acc};
+            (_, or_, Acc) ->
+                {skip, Acc};
+            (Conditional, Filter, Acc) ->
+                check_where_clause_is_possible_fold(DDL, Conditional, Filter, Acc)
+        end, [{eliminate_later,[]}], Filter1),
+    ElimLater = proplists:get_value(eliminate_later, FilterChecks),
+    {Filter3, _} = riak_ql_ddl:mapfold_where_tree(
+        fun (_, and_, Acc) ->
+                {ok, Acc};
+            (_, or_, Acc) ->
+                {skip, Acc};
+            (_, Filter, Acc) ->
+                case lists:member(Filter, ElimLater) of
+                  true ->
+                      {eliminate, Acc};
+                  false ->
+                      {Filter, Acc}
+                end
+        end, [], Filter2),
+    lists:keystore(filter, 1, WhereProps, {filter,Filter3}).
+
+check_where_clause_is_possible_fold(DDL, _, {'=',{field,FieldName,_},{const,?SQL_NULL}} = F, Acc) ->
+    %% `IS NULL` filters
+    %% if a column is marked NOT NULL, but the WHERE clause has IS NULL for that
+    %% column then no results will ever be returned, so do not execute!
+    case is_field_nullable(FieldName, DDL) of
+        true ->
+            {F, Acc};
+        false ->
+            throw({error, {impossible_where_clause, << >>}})
+    end;
+check_where_clause_is_possible_fold(DDL, _, {'!=',{field,FieldName,_},{const,?SQL_NULL}} = F, Acc) ->
+    %% `NOT NULL` filters
+    %% if column is marked NOT NULL, and the WHERE clause has IS NOT NULL for
+    %% that column then we can eliminate it, and reduce the filter, maybe to
+    %% nothing which means leveldb would not have to decode the rows!
+    case is_field_nullable(FieldName, DDL) of
+        true ->
+            {F, Acc};
+        false ->
+            {eliminate, Acc}
+    end;
+check_where_clause_is_possible_fold(_, _, {'=',{field,FieldName,_},{const,EqVal}} = F, Acc) ->
+    case find_filter_check(FieldName, Acc) of
+        #filtercheck{'=' = F} ->
+            %% this filter has been specified twice so remove the second occurence
+            {eliminate, Acc};
+        #filtercheck{'>' = {_,_,{const,GtVal}} = GtF} = Check1 when GtVal < EqVal ->
+            %% query like `a = 10 AND a > 8` we can eliminate the greater than
+            %% clause because if a to be 10 it must also be greater than 8
+            Acc2 = append_to_eliminate_later(GtF, Acc),
+            Check2 = Check1#filtercheck{'>' = undefined, '=' = F},
+            {F, lists:keystore(FieldName, #filtercheck.name, Acc2, Check2)};
+        #filtercheck{'>=' = {_,_,{const,GteVal}} = GteF} = Check1 when GteVal < EqVal ->
```

Comment:
'>' and '>=' and '<' and '<=' cases that differ only in the operator's inclusion of '=' should be DRY'd up using guards b/c there's more than a simple line of code w/i the case's statement(s) list.